Source/DafnyCore/Compilers/Rust/Dafny-compiler-rust-proofs.dfy

```diff
@@ -0,0 +1,362 @@
+  }
+
+  lemma {:rlimit 200} idiomatic_rust_reversible(i: string)
+    requires !has_special(i)
+    ensures reverse_idiomatic_rust(idiomatic_rust(i)) == i
+  {
+    if |i| == 0 {
+    } else if i[0] == '_' {
+      idiomatic_rust_reversible(i[2..]);
+      assert 2 <= |i| && i[1] == '_';
+      var ss := ['_'] + idiomatic_rust(i[2..]);
+      assert ss[0] == '_' && |ss| != 0;
+    } else {
+      idiomatic_rust_reversible(i[1..]);
+    }
+  }
+
+  function reverseEscapeIdent(s: string): string {
+    if |s| >= 2 && s[0..2] == "r#" then
+      if |s| >= 3 && s[2] == '_' then
+        reverseReplaceDots(s[3..]) // General escape
+      else
+        s[2..] // Keyword
+    else if |s| >= 1 && s[0] == '_' then
+      if |s| >= 2 && s[1] == 'T' then
+        "___hMake" + s[2..] // Tuple builder
+      else if is_tuple_numeric(s) then
+        s
+      else
+        s
+    else // Idiomatic rust
+      reverse_idiomatic_rust(s)
+  }
+
+  lemma tupleIdentReversible(s: string)
+    requires is_tuple_numeric(s)
+    ensures reverseEscapeIdent(escapeIdent(s)) == s
+  {}
+
+  lemma {:rlimit 500} tupleBuilderReversible(i: string)
+    requires !is_tuple_numeric(i)
+    requires is_tuple_builder(i)
+    ensures reverseEscapeIdent(escapeIdent(i)) == i
+  {
+    var s := "_T" + i[8..];
+    calc {
+      reverseEscapeIdent(escapeIdent(i));
+      reverseEscapeIdent(s);
+      i[0..8] + s[2..];
+      i;
+    }
+  }
+
+  lemma {:rlimit 800} reservedRustReversible(i: string)
+    requires !is_tuple_numeric(i)
+    requires !is_tuple_builder(i)
+    requires i in reserved_rust
+    ensures reverseEscapeIdent(escapeIdent(i)) == i
+  {
+    assert i[0] != '_';
+    var s := "r#" + i;
+    calc {
+      reverseEscapeIdent(escapeIdent(i));
+      reverseEscapeIdent(s);
+      s[2..];
+      i;
+    }
+  }
+
+  lemma {:rlimit 800} idiomaticRustReversible(i: string)
+    requires isDafnyEncodedId(i) // Needed to ensure i does not start with '_'
+    requires !is_tuple_numeric(i)
+    requires !is_tuple_builder(i)
+    requires i !in reserved_rust
+    requires is_idiomatic_rust(i)
+    ensures reverseEscapeIdent(escapeIdent(i)) == i
+  {
+    var s := idiomatic_rust(i);
+    idiomatic_rust_not_rsharp(i);
+    assert |s| < 2 || s[0..2] != "r#" by {
+      if |s| >= 2 && s[0..2] == "r#" {
+        assert s[1] == '#';
+        idiomatic_rust_not_rsharp(i[1..]);
+      }
+    }
+    assert s[0] != '_';
+    calc {
+      reverseEscapeIdent(escapeIdent(i));
+      reverseEscapeIdent(idiomatic_rust(i));
+      reverse_idiomatic_rust(idiomatic_rust(i));
+      { idiomatic_rust_reversible(i); }
+      i;
+    }
+  }
+
+  lemma {:rlimit 800} dafnyGeneratedIdReversible(s: string)
+    requires !is_tuple_numeric(s)
+    requires !is_tuple_builder(s)
+    requires s !in reserved_rust
+    requires !is_idiomatic_rust(s)
+    requires is_dafny_generated_id(s)
+    ensures reverseEscapeIdent(escapeIdent(s)) == s
+  {
+  }
+
+  lemma {:rlimit 800} everythingElseReversible(i: string)
```

Comment:
Either "Unescapable" or "Invertable" would be better than "Reversible" in general IMO

Reply:
I replaced every "Reversible" by "Invertible" in this file.